crates/lib/src/name.rs

```diff
@@ -287,69 +266,56 @@ impl From<Tld> for DomainPart {
 /// Each segment in a database name can contain any UTF-8 character, except for
 /// whitespace and '/'. The maximum segment length is 64 characters.
 ///
-/// [`DomainName`]s consist of [`DomainPart`]s, and are compared case-
-/// insensitively. Note, however, that the [`fmt::Display`] and [`Serialize`]
-/// impls will use the original, (potentially) mixed-case representation.
+/// Note that [`PartialEq`] compares the exact string representation of a
+/// [`DomainName`], as one would expect, but the SpacetimeDB registry compares
+/// the lowercase representation of it.
 ///
 /// To construct a valid [`DomainName`], use [`parse_domain_name`] or the
 /// [`FromStr`] impl.
-#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[derive(Debug, Clone, PartialEq, Eq)]
 pub struct DomainName {
-    /// The top level domain for the domain name. For example:
-    ///
-    ///  * `clockworklabs/bitcraft`
-    ///
-    /// Here, `clockworklabs` is the tld.
-    tld: Tld,
-    /// The part after the top level domain, this is not required. For example:
-    ///
-    ///  * `clockworklabs/bitcraft`
-    ///
-    /// Here, `bitcraft` is the subdomain.
-    sub_domain: Option<DomainPart>,
+    // Iff there is a subdomain, next char in `domain_name` is '/'.
+    tld_offset: usize,
+    domain_name: String,
 }
 
 impl DomainName {
-    pub fn tld(&self) -> &Tld {
-        &self.tld
+    pub fn as_str(&self) -> &str {
+        &self.domain_name
     }
 
-    /// Drop subdomain, if any, and return only the TLD
-    pub fn into_tld(self) -> Tld {
-        self.tld
+    pub fn tld(&self) -> &TldRef {
+        TldRef::new(&self.domain_name[..self.tld_offset])
     }
 
-    pub fn sub_domain(&self) -> Option<&DomainPart> {
-        self.sub_domain.as_ref()
+    pub fn as_tld(&self) -> Tld {
```

Comment:
I beg to differ: analogous to the `Into` trait, `into_*` inherents usually consume the receiver (which is not the case here). The difference between `as_*` and `to_*` is often less clear, but `to_` usually suggests that allocation may happen. Since that's the case here, I'd be fine with `to_tld`.